# Credit Card Approval Predition  
appying  Binary classification Problem  
 Using some  machine learning technique to predit weather a persons application for credit card  should be accepted or rejected

In [ ]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('/content/Credit_card.csv')
df

Ind_ID GENDER Car_Owner Propert_Owner  CHILDREN  Annual_income  \
0     5008827      M         Y             Y         0       180000.0   
1     5009744      F         Y             N         0       315000.0   
2     5009746      F         Y             N         0       315000.0   
3     5009749      F         Y             N         0            NaN   
4     5009752      F         Y             N         0       315000.0   
...       ...    ...       ...           ...       ...            ...   
1494  5009000      M         Y             Y         0       157500.0   
1495  5090302      F         N             Y         0       405000.0   
1496  5067299      M         N             Y         0       225000.0   
1497  5010344      F         N             N         0       216000.0   
1498  5011032      F         N             Y         1       135000.0   

               Type_Income                      EDUCATION  \
0                Pensioner               Higher education   
1     Commercial associate               Higher education   
2     Commercial associate               Higher education   
3     Commercial associate               Higher education   
4     Commercial associate               Higher education   
...                    ...                            ...   
1494               Working  Secondary / secondary special   
1495  Commercial associate  Secondary / secondary special   
1496  Commercial associate  Secondary / secondary special   
1497             Pensioner  Secondary / secondary special   
1498               Working                Academic degree   

            Marital_status       Housing_type  Birthday_count  Employed_days  \
0                  Married  House / apartment        -18772.0         365243   
1                  Married  House / apartment        -13557.0           -586   
2                  Married  House / apartment             NaN           -586   
3                  Married  House / apartment        -13557.0           -586   
4                  Married  House / apartment        -13557.0           -586   
...                    ...                ...             ...            ...   
1494               Married  House / apartment         -9927.0           -828   
1495               Married  House / apartment        -17427.0          -4874   
1496  Single / not married  House / apartment         -9149.0          -1217   
1497               Married  House / apartment        -22867.0         365243   
1498        Civil marriage  House / apartment        -17489.0          -7557   

      Mobile_phone  Work_Phone  Phone  EMAIL_ID Type_Occupation  \
0                1           0      0         0             NaN   
1                1           1      1         0             NaN   
2                1           1      1         0             NaN   
3                1           1      1         0             NaN   
4                1           1      1         0             NaN   
...            ...         ...    ...       ...             ...   
1494             1           0      0         0         Drivers   
1495             1           0      0         1     Sales staff   
1496             1           0      0         1        Laborers   
1497             1           0      0         0             NaN   
1498             1           0      1         0             NaN   

      Family_Members  label  
0                  2      1  
1                  2      1  
2                  2      1  
3                  2      1  
4                  2      1  
...              ...    ...  
1494               2      0  
1495               2      0  
1496               1      0  
1497               2      0  
1498               3      0  

[1499 rows x 19 columns]

Explaing the  information provided in each coloumns  
Features name: (Credit_Card.csv)

1) Ind_ID: Client Id

2) Gender: Gender information

3) Car_owner: Having car or not

4) Propert_owner: Having property or not

5) Children: Count of children

6) Annual_income: Annual income

7) Type_Income: Income type

8) Education: Education level

9) Marital_status: Marital status

10) Housing_type: Living style

11) Birthday_count: Use backward count from current day (0), -1 means yesterday.

12) Employed_days: Start date of employment. Use backward count from current day (0). Positive value means, individual is currently unemployed.

13) Mobile_phone: Any mobile phone

14) Work_phone: Any work phone

15) Phone: Any phone number

16) EMAIL_ID: Any email ID

17) Type_Occupation: Occupation

18) Family_Members: Family size

19 )Label: 0 is application approved and 1 is application rejected.










In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1499 entries, 0 to 1498
Data columns (total 19 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Ind_ID           1499 non-null   int64  
 1   GENDER           1492 non-null   object 
 2   Car_Owner        1499 non-null   object 
 3   Propert_Owner    1499 non-null   object 
 4   CHILDREN         1499 non-null   int64  
 5   Annual_income    1477 non-null   float64
 6   Type_Income      1499 non-null   object 
 7   EDUCATION        1499 non-null   object 
 8   Marital_status   1499 non-null   object 
 9   Housing_type     1499 non-null   object 
 10  Birthday_count   1478 non-null   float64
 11  Employed_days    1499 non-null   int64  
 12  Mobile_phone     1499 non-null   int64  
 13  Work_Phone       1499 non-null   int64  
 14  Phone            1499 non-null   int64  
 15  EMAIL_ID         1499 non-null   int64  
 16  Type_Occupation  1028 non-null   object 
 17  Family_Members

In [ ]:
df.isna().sum().sum()

521

result_df = pd.DataFrame(columns=['Column', 'Null Count', 'Unique Values'])

# Iterate over each column in the DataFrame
for column in df.columns:
    null_count = df[column].isnull().sum()  # Count null values
    unique_values = df[column].nunique()  # Count unique values
    unique_terms = df[column].unique()  # Get unique terms
    result_df = result_df.append({'Column': column,
                                  'Null Count': null_count,
                                  'Unique Values': unique_values,},ignore_index=True
                                  )

# Print the result DataFrame
print(result_df)

In [6]:
#result_df = pd.DataFrame(columns=['Column', 'Null Count', 'Unique Values'])

# Iterate over each column in the DataFrame
for column in df.columns:
    null_count = df[column].isnull().sum()  # Count null values
    unique_values = df[column].nunique()  # Count unique values
    unique_terms = df[column].unique()  # Get unique terms
    df1=pd.DataFrame({'Column': [column],
                    'Null Count': [null_count],
                  'Unique Values': [unique_values]})


# Print the result DataFrame
print(df1)

  Column  Null Count  Unique Values
0  label           0              2


In [10]:
# Add a column 'Unemployed' and initialize with 0
df['Unemployed'] = 0

# Make negative values in 'Employed_days' positive
df['Employed_days'] = df['Employed_days'].abs()

# Move positive values from 'Employed_days' to 'Unemployed'
positive_values = df['Employed_days'][df['Employed_days'] > 0]
df.loc[df['Employed_days'] > 0, 'Unemployed'] = positive_values

# Set all positive values in 'Employed_days' to 0
df.loc[df['Employed_days'] > 0, 'Employed_days'] = 0

df

Ind_ID GENDER Car_Owner Propert_Owner  CHILDREN  Annual_income  \
0     5008827      M         Y             Y         0       180000.0   
1     5009744      F         Y             N         0       315000.0   
2     5009746      F         Y             N         0       315000.0   
3     5009749      F         Y             N         0            NaN   
4     5009752      F         Y             N         0       315000.0   
...       ...    ...       ...           ...       ...            ...   
1494  5009000      M         Y             Y         0       157500.0   
1495  5090302      F         N             Y         0       405000.0   
1496  5067299      M         N             Y         0       225000.0   
1497  5010344      F         N             N         0       216000.0   
1498  5011032      F         N             Y         1       135000.0   

               Type_Income                      EDUCATION  \
0                Pensioner               Higher education   
1     Commercial associate               Higher education   
2     Commercial associate               Higher education   
3     Commercial associate               Higher education   
4     Commercial associate               Higher education   
...                    ...                            ...   
1494               Working  Secondary / secondary special   
1495  Commercial associate  Secondary / secondary special   
1496  Commercial associate  Secondary / secondary special   
1497             Pensioner  Secondary / secondary special   
1498               Working                Academic degree   

            Marital_status       Housing_type  Birthday_count  Employed_days  \
0                  Married  House / apartment        -18772.0              0   
1                  Married  House / apartment        -13557.0              0   
2                  Married  House / apartment             NaN              0   
3                  Married  House / apartment        -13557.0              0   
4                  Married  House / apartment        -13557.0              0   
...                    ...                ...             ...            ...   
1494               Married  House / apartment         -9927.0              0   
1495               Married  House / apartment        -17427.0              0   
1496  Single / not married  House / apartment         -9149.0              0   
1497               Married  House / apartment        -22867.0              0   
1498        Civil marriage  House / apartment        -17489.0              0   

      Mobile_phone  Work_Phone  Phone  EMAIL_ID Type_Occupation  \
0                1           0      0         0             NaN   
1                1           1      1         0             NaN   
2                1           1      1         0             NaN   
3                1           1      1         0             NaN   
4                1           1      1         0             NaN   
...            ...         ...    ...       ...             ...   
1494             1           0      0         0         Drivers   
1495             1           0      0         1     Sales staff   
1496             1           0      0         1        Laborers   
1497             1           0      0         0             NaN   
1498             1           0      1         0             NaN   

      Family_Members  label  Unemployed  
0                  2      1           0  
1                  2      1           0  
2                  2      1           0  
3                  2      1           0  
4                  2      1           0  
...              ...    ...         ...  
1494               2      0           0  
1495               2      0           0  
1496               1      0           0  
1497               2      0           0  
1498               3      0           0  

[1499 rows x 20 columns]

In [ ]:
null_income_rows = df[df['Annual_income'].isnull()]

# Print rows with null values in 'Annual_income' column
print("Rows with null values in 'Annual_income' column:")
print(null_income_rows)

Rows with null values in 'Annual_income' column:
       Ind_ID GENDER Car_Owner Propert_Owner  CHILDREN  Annual_income  \
3     5009749      F         Y             N         0            NaN   
12    5018501      F         Y             Y         0            NaN   
27    5024917      F         N             Y         0            NaN   
65    5058348      F         N             Y         0            NaN   
76    5062137      M         N             Y         1            NaN   
107   5105056      M         Y             N         0            NaN   
125   5116305      F         N             Y         0            NaN   
152   5142183      M         Y             Y         2            NaN   
185   5033574      F         Y             Y         1            NaN   
215   5085769      F         N             Y         0            NaN   
273   5047704      F         N             Y         2            NaN   
324   5069227      F         Y             Y         0            NaN   
38

In [ ]:
# Group data by 'Type_Occupation' and calculate the mean of 'Annual_income'
occupation_mean_income = df.groupby('Type_Income')['Annual_income'].mean()

# Print unique values of 'Type_Occupation' and their mean annual incomes
print("Unique values of 'Type_Occupation' and their mean annual incomes:")
for occupation, mean_income in occupation_mean_income.items():
    print(f"{occupation}: {mean_income}")


Unique values of 'Type_Occupation' and their mean annual incomes:
Commercial associate: 236101.7195467422
Pensioner: 155275.2096153846
State servant: 212142.85714285713
Working: 182596.23126614987


In [ ]:
occupation_mean_income_dict = {
    'Commercial associate': 236101,
    'Pensioner': 154012,
    'State servant': 212142,
    'Working': 181824
}

# Fill null values in 'Annual_income' based on 'Type_Occupation'
for occupation, mean_income in occupation_mean_income_dict.items():
    df['Annual_income'].fillna(mean_income, inplace=True)

# Verify the DataFrame after filling null values
df


Ind_ID GENDER Car_Owner Propert_Owner  CHILDREN  Annual_income  \
0     5008827      M         Y             Y         0       180000.0   
1     5009744      F         Y             N         0       315000.0   
2     5009746      F         Y             N         0       315000.0   
3     5009749      F         Y             N         0       236101.0   
4     5009752      F         Y             N         0       315000.0   
...       ...    ...       ...           ...       ...            ...   
1494  5009000      M         Y             Y         0       157500.0   
1495  5090302      F         N             Y         0       405000.0   
1496  5067299      M         N             Y         0       225000.0   
1497  5010344      F         N             N         0       216000.0   
1498  5011032      F         N             Y         1       135000.0   

               Type_Income                      EDUCATION  \
0                Pensioner               Higher education   
1     Commercial associate               Higher education   
2     Commercial associate               Higher education   
3     Commercial associate               Higher education   
4     Commercial associate               Higher education   
...                    ...                            ...   
1494               Working  Secondary / secondary special   
1495  Commercial associate  Secondary / secondary special   
1496  Commercial associate  Secondary / secondary special   
1497             Pensioner  Secondary / secondary special   
1498               Working                Academic degree   

            Marital_status       Housing_type  Birthday_count  Employed_days  \
0                  Married  House / apartment        -18772.0              0   
1                  Married  House / apartment        -13557.0              0   
2                  Married  House / apartment             NaN              0   
3                  Married  House / apartment        -13557.0              0   
4                  Married  House / apartment        -13557.0              0   
...                    ...                ...             ...            ...   
1494               Married  House / apartment         -9927.0              0   
1495               Married  House / apartment        -17427.0              0   
1496  Single / not married  House / apartment         -9149.0              0   
1497               Married  House / apartment        -22867.0              0   
1498        Civil marriage  House / apartment        -17489.0              0   

      Mobile_phone  Work_Phone  Phone  EMAIL_ID Type_Occupation  \
0                1           0      0         0             NaN   
1                1           1      1         0             NaN   
2                1           1      1         0             NaN   
3                1           1      1         0             NaN   
4                1           1      1         0             NaN   
...            ...         ...    ...       ...             ...   
1494             1           0      0         0         Drivers   
1495             1           0      0         1     Sales staff   
1496             1           0      0         1        Laborers   
1497             1           0      0         0             NaN   
1498             1           0      1         0             NaN   

      Family_Members  label  Unemployed  
0                  2      1           0  
1                  2      1           0  
2                  2      1           0  
3                  2      1           0  
4                  2      1           0  
...              ...    ...         ...  
1494               2      0           0  
1495               2      0           0  
1496               1      0           0  
1497               2      0           0  
1498               3      0           0  

[1499 rows x 20 columns]

In [ ]:
df['Type_Occupation'].fillna('Not Mentioned',inplace=True)

In [ ]:
df.isna().sum()

Ind_ID              0
GENDER              7
Car_Owner           0
Propert_Owner       0
CHILDREN            0
Annual_income       0
Type_Income         0
EDUCATION           0
Marital_status      0
Housing_type        0
Birthday_count     21
Employed_days       0
Mobile_phone        0
Work_Phone          0
Phone               0
EMAIL_ID            0
Type_Occupation     0
Family_Members      0
label               0
Unemployed          0
dtype: int64

In [ ]:
import matplotlib.pyplot as plt
from sklearn.preprocessing import LabelEncoder


In [ ]:
# Initialize the LabelEncoder
label_encoder = LabelEncoder()
# Fit and transform the 'Size' column
df['Car_Owner(encoded)'] = label_encoder.fit_transform(df['Car_Owner'])
df['Propert_Owner(encoded)'] = label_encoder.fit_transform(df['Propert_Owner'])
df['Type_Income(encoded)'] = label_encoder.fit_transform(df['Type_Income'])
df['EDUCATION(encoded)']  = label_encoder.fit_transform(df['EDUCATION'])
df['Housing_type(encoded)']  = label_encoder.fit_transform(df['Housing_type'])
df['Marital_status(encoded)']=label_encoder.fit_transform(df['Marital_status'])
df['Type_Occupation(encoded)']=label_encoder.fit_transform(df['Type_Occupation'])


In [ ]:
from sklearn.model_selection import train_test_split

# Splitting the dataset into features (X) and target (y)
X = df[['Car_Owner(encoded)', 'Propert_Owner(encoded)', 'CHILDREN', 'Annual_income',
        'Type_Income(encoded)', 'EDUCATION(encoded)', 'Marital_status(encoding)', 'Housing_type(encoded)',
         'Employed_days', 'Mobile_phone', 'Work_Phone',
        'Phone', 'EMAIL_ID', 'Type_Occupation(encoding)', 'Family_Members']]
y = df['label']

# Splitting the dataset into train and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=100)


In [ ]:
# Create a mask to identify rows with negative values in the training data
negative_mask = (X_train < 0).any(axis=1)

# Filter the training data to include only rows with negative values
rows_with_negative_values = X_train[negative_mask]

# Print the rows containing negative values
print("Rows containing negative values in the training data:")
print(rows_with_negative_values)


Rows containing negative values in the training data:
Empty DataFrame
Columns: [Car_Owner(encoded), Propert_Owner(encoded), CHILDREN, Annual_income, Type_Income(encoded), EDUCATION(encoded), Marital_status(encoding), Housing_type(encoded), Employed_days, Mobile_phone, Work_Phone, Phone, EMAIL_ID, Type_Occupation(encoding), Family_Members]
Index: []


ValueError                                Traceback (most recent call last)
<ipython-input-72-4bf6057fa576> in <cell line: 2>()
      1 from sklearn.feature_selection import chi2
----> 2 f_p_values=chi2(X_train,y_train)

/usr/local/lib/python3.10/dist-packages/sklearn/feature_selection/_univariate_selection.py in chi2(X, y)
    215     X = check_array(X, accept_sparse="csr", dtype=(np.float64, np.float32))
    216     if np.any((X.data if issparse(X) else X) < 0):
--> 217         raise ValueError("Input X must be non-negative.")
    218
    219     # Use a sparse representation for Y by default to reduce memory usage when

ValueError: Input X must be non-negative.


In [ ]:
from sklearn.feature_selection import chi2
f_p_values=chi2(X_train,y_train)

In [ ]:
p_values=pd.Series(f_p_values[1])
p_values.index=X_train.columns
p_values

Car_Owner(encoded)           0.979933
Propert_Owner(encoded)       0.461060
CHILDREN                     0.536851
Annual_income                0.000000
Type_Income(encoded)         0.026676
EDUCATION(encoded)           0.522491
Marital_status(encoding)     0.127178
Housing_type(encoded)        0.984411
Employed_days                     NaN
Mobile_phone                 1.000000
Work_Phone                   0.936985
Phone                        0.832981
EMAIL_ID                     0.770603
Type_Occupation(encoding)    0.711850
Family_Members               0.472782
dtype: float64